File: melvin_brain_monitor.py
#!/usr/bin/env python3
"""
🧠 MELVIN BRAIN MONITOR - Real-time Neural Activity Tracking
===========================================================
Monitor Melvin's brain activity, node creation, connection formation, and learning patterns.
Track Hebbian learning, pruning decisions, and neural network growth in real-time.
"""

import time
import json
import threading
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Optional
from collections import defaultdict, deque

from melvin_optimized_v2 import MelvinOptimizedV2, OptimizedContinuousFeeder, ContentType, ConnectionType

class MelvinBrainMonitor:
    """Real-time monitoring system for Melvin's brain activity"""
    
    def __init__(self, melvin_brain: MelvinOptimizedV2):
        self.melvin = melvin_brain
        self.monitoring = False
        self.monitor_thread = None
        
        # Activity tracking
        self.activity_log = deque(maxlen=10000)
        self.node_creation_times = []
        self.connection_formation_times = []
        self.hebbian_events = []
        self.pruning_events = []
        
        # Real-time statistics
        self.stats_history = deque(maxlen=1000)
        self.connection_types = defaultdict(int)
        self.content_types = defaultdict(int)
        self.activation_patterns = defaultdict(int)
        
        # Learning analysis
        self.learning_rate = []
        self.importance_scores = []
        self.compression_ratios = []
        
        # Monitoring configuration
        self.update_interval = 1.0  # seconds
        self.detailed_logging = True
        
        print("🧠 Melvin Brain Monitor initialized")
    
    def start_monitoring(self):
        """Start real-time brain monitoring"""
        self.monitoring = True
        self.monitor_thread = threading.Thread(target=self._monitor_loop, daemon=True)
        self.monitor_thread.start()
        print("📊 Brain monitoring started - tracking neural activity...")
    
    def stop_monitoring(self):
        """Stop brain monitoring"""
        self.monitoring = False
        if self.monitor_thread:
            self.monitor_thread.join()
        print("📊 Brain monitoring stopped")
    
    def _monitor_loop(self):
        """Main monitoring loop"""
        while self.monitoring:
            try:
                # Capture current brain state
                state = self.melvin.get_unified_state()
                self._record_activity(state)
                
                # Update statistics
                self._update_stats(state)
                
                # Log detailed activity
                if self.detailed_logging:
                    self._log_detailed_activity(state)
                
                time.sleep(self.update_interval)
                
            except Exception as e:
                print(f"⚠️ Monitoring error: {e}")
                time.sleep(self.update_interval)
    
    def _record_activity(self, state: Dict[str, Any]):
        """Record brain activity"""
        timestamp = time.time()
        
        activity = {
            'timestamp': timestamp,
            'datetime': datetime.fromtimestamp(timestamp).isoformat(),
            'nodes': state['global_memory']['total_nodes'],
            'connections': state['global_memory']['total_edges'],
            'storage_mb': state['global_memory']['storage_used_mb'],
            'hebbian_updates': state['global_memory']['stats']['hebbian_updates'],
            'uptime_seconds': state['system']['uptime_seconds']
        }
        
        self.activity_log.append(activity)
        self.stats_history.append(activity)
    
    def _update_stats(self, state: Dict[str, Any]):
        """Update real-time statistics"""
        # Track connection types
        if hasattr(self.melvin.binary_storage, 'total_connections'):
            # This would need to be enhanced to track actual connection types
            pass
        
        # Track content types
        if hasattr(self.melvin.binary_storage, 'total_nodes'):
            # This would need to be enhanced to track actual content types
            pass
    
    def _log_detailed_activity(self, state: Dict[str, Any]):
        """Log detailed brain activity"""
        current_time = datetime.now().strftime("%H:%M:%S")
        
        print(f"\n🧠 [{current_time}] BRAIN ACTIVITY UPDATE:")
        print(f"   📦 Nodes: {state['global_memory']['total_nodes']:,}")
        print(f"   🔗 Connections: {state['global_memory']['total_edges']:,}")
        print(f"   💾 Storage: {state['global_memory']['storage_used_mb']:.2f}MB")
        print(f"   ⚡ Hebbian Updates: {state['global_memory']['stats']['hebbian_updates']}")
        print(f"   🕐 Uptime: {state['system']['uptime_seconds']:.1f}s")
        
        # Calculate growth rates
        if len(self.stats_history) > 1:
            prev = self.stats_history[-2]
            curr = self.stats_history[-1]
            
            node_growth = curr['nodes'] - prev['nodes']
            connection_growth = curr['connections'] - prev['connections']
            
            if node_growth > 0:
                print(f"   📈 Node Growth: +{node_growth}")
            if connection_growth > 0:
                print(f"   🔗 Connection Growth: +{connection_growth}")
    
    def feed_data_and_monitor(self, data_sources: List[str], duration: Optional[int] = None):
        """Feed data to Melvin while monitoring brain activity"""
        print(f"🚀 Starting data feeding with brain monitoring...")
        print(f"📖 Data sources: {len(data_sources)}")
        print(f"⏱️ Duration: {'Continuous' if duration is None else f'{duration}s'}")
        
        # Start monitoring
        self.start_monitoring()
        
        # Create feeder
        feeder = OptimizedContinuousFeeder(max_storage_gb=4000)
        
        start_time = time.time()
        try:
            # Process data sources
            for i, source_path in enumerate(data_sources):
                if duration and (time.time() - start_time) > duration:
                    print(f"⏰ Time limit reached ({duration}s)")
                    break
                
                print(f"\n📖 Processing source {i+1}/{len(data_sources)}: {source_path}")
                
                # Extract and process content
                content_list = feeder._extract_content_from_source(source_path)
                
                if not content_list:
                    print(f"⚠️ No content extracted from {source_path}")
                    continue
                
                # Process each content item with monitoring
                for j, content in enumerate(content_list):
                    if duration and (time.time() - start_time) > duration:
                        break
                    
                    if isinstance(content, str):
                        # Process text input
                        node_id = self.melvin.process_text_input(content, "monitor_feeder")
                        
                        # Record node creation
                        self.node_creation_times.append({
                            'timestamp': time.time(),
                            'node_id': node_id.hex()[:8],
                            'content_type': 'TEXT',
                            'content_preview': content[:50] + "..." if len(content) > 50 else content
                        })
                        
                        # Monitor Hebbian learning
                        if hasattr(self.melvin, 'recent_activations'):
                            recent_count = len(self.melvin.recent_activations)
                            if recent_count > 1:
                                self.hebbian_events.append({
                                    'timestamp': time.time(),
                                    'node_id': node_id.hex()[:8],
                                    'co_activations': recent_count - 1
                                })
                    
                    # Progress update
                    if j % 10 == 0:
                        state = self.melvin.get_unified_state()
                        print(f"   📊 Progress: {j+1}/{len(content_list)} items, "
                              f"{state['global_memory']['total_nodes']} nodes, "
                              f"{state['global_memory']['total_edges']} connections")
                
                print(f"✅ Completed {source_path}")
                
        except KeyboardInterrupt:
            print("\n🛑 Data feeding interrupted by user")
        finally:
            # Stop monitoring
            self.stop_monitoring()
            
            # Final report
            self._generate_final_report()
    
    def _generate_final_report(self):
        """Generate comprehensive brain activity report"""
        print("\n" + "="*60)
        print("🧠 MELVIN BRAIN ACTIVITY REPORT")
        print("="*60)
        
        if not self.stats_history:
            print("❌ No activity data recorded")
            return
        
        # Calculate statistics
        initial_state = self.stats_history[0]
        final_state = self.stats_history[-1]
        
        # Growth statistics
        total_nodes_created = final_state['nodes'] - initial_state['nodes']
        total_connections_formed = final_state['connections'] - initial_state['connections']
        total_hebbian_updates = final_state['hebbian_updates'] - initial_state['hebbian_updates']
        
        # Time statistics
        total_time = final_state['uptime_seconds'] - initial_state['uptime_seconds']
        
        # Rates
        nodes_per_second = total_nodes_created / total_time if total_time > 0 else 0
        connections_per_second = total_connections_formed / total_time if total_time > 0 else 0
        hebbian_per_second = total_hebbian_updates / total_time if total_time > 0 else 0
        
        print(f"\n📊 GROWTH STATISTICS:")
        print(f"   🧠 Nodes Created: {total_nodes_created:,}")
        print(f"   🔗 Connections Formed: {total_connections_formed:,}")
        print(f"   ⚡ Hebbian Updates: {total_hebbian_updates:,}")
        print(f"   💾 Storage Used: {final_state['storage_mb']:.2f}MB")
        
        print(f"\n⏱️ TIME STATISTICS:")
        print(f"   🕐 Total Time: {total_time:.1f}s")
        print(f"   📈 Nodes/Second: {nodes_per_second:.2f}")
        print(f"   🔗 Connections/Second: {connections_per_second:.2f}")
        print(f"   ⚡ Hebbian/Second: {hebbian_per_second:.2f}")
        
        print(f"\n🎯 LEARNING PATTERNS:")
        print(f"   🧠 Node Creation Events: {len(self.node_creation_times)}")
        print(f"   🔗 Hebbian Events: {len(self.hebbian_events)}")
        print(f"   📊 Activity Logs: {len(self.activity_log)}")
        
        # Connection formation analysis
        if self.hebbian_events:
            avg_co_activations = sum(e['co_activations'] for e in self.hebbian_events) / len(self.hebbian_events)
            print(f"   🔗 Avg Co-activations: {avg_co_activations:.2f}")
        
        # Storage efficiency
        if total_nodes_created > 0:
            avg_node_size = final_state['storage_mb'] * 1024 * 1024 / final_state['nodes']
            print(f"   💾 Avg Node Size: {avg_node_size:.1f} bytes")
        
        print(f"\n📈 FINAL STATE:")
        print(f"   🧠 Total Nodes: {final_state['nodes']:,}")
        print(f"   🔗 Total Connections: {final_state['connections']:,}")
        print(f"   💾 Total Storage: {final_state['storage_mb']:.2f}MB")
        print(f"   ⚡ Total Hebbian: {final_state['hebbian_updates']:,}")
        
        # Save detailed report
        self._save_detailed_report()
    
    def _save_detailed_report(self):
        """Save detailed brain activity report to file"""
        report = {
            'timestamp': datetime.now().isoformat(),
            'summary': {
                'total_nodes': self.stats_history[-1]['nodes'] if self.stats_history else 0,
                'total_connections': self.stats_history[-1]['connections'] if self.stats_history else 0,
                'total_storage_mb': self.stats_history[-1]['storage_mb'] if self.stats_history else 0,
                'total_hebbian': self.stats_history[-1]['hebbian_updates'] if self.stats_history else 0
            },
            'activity_log': list(self.activity_log),
            'node_creation_times': self.node_creation_times,
            'hebbian_events': self.hebbian_events,
            'stats_history': list(self.stats_history)
        }
        
        # Save to file
        report_file = f"melvin_brain_report_{int(time.time())}.json"
        with open(report_file, 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"\n💾 Detailed report saved to: {report_file}")
    
    def get_real_time_stats(self) -> Dict[str, Any]:
        """Get current real-time statistics"""
        if not self.stats_history:
            return {}
        
        current = self.stats_history[-1]
        
        # Calculate rates
        if len(self.stats_history) > 1:
            prev = self.stats_history[-2]
            time_diff = current['timestamp'] - prev['timestamp']
            
            node_rate = (current['nodes'] - prev['nodes']) / time_diff if time_diff > 0 else 0
            connection_rate = (current['connections'] - prev['connections']) / time_diff if time_diff > 0 else 0
        else:
            node_rate = 0
            connection_rate = 0
        
        return {
            'current_nodes': current['nodes'],
            'current_connections': current['connections'],
            'current_storage_mb': current['storage_mb'],
            'current_hebbian': current['hebbian_updates'],
            'node_rate_per_second': node_rate,
            'connection_rate_per_second': connection_rate,
            'uptime_seconds': current['uptime_seconds']
        }

def main():
    """Main function for brain monitoring demonstration"""
    print("🧠 MELVIN BRAIN MONITOR")
    print("=" * 50)
    
    # Initialize Melvin's optimized brain
    melvin = MelvinOptimizedV2()
    
    # Initialize brain monitor
    monitor = MelvinBrainMonitor(melvin)
    
    # Define data sources to feed
    data_sources = [
        "README.md",
        "melvin_global_brain.py",
        "melvin_optimized_v2.py",
        "MELVIN_OPTIMIZED_V2_SUMMARY.md"
    ]
    
    # Start feeding data with monitoring
    print("🚀 Starting brain activity monitoring...")
    print("Press Ctrl+C to stop monitoring")
    
    try:
        # Feed data and monitor brain activity
        monitor.feed_data_and_monitor(data_sources, duration=60)  # Monitor for 60 seconds
        
    except KeyboardInterrupt:
        print("\n🛑 Monitoring stopped by user")
        monitor.stop_monitoring()
        monitor._generate_final_report()
    
    print("\n🎉 Brain monitoring session completed!")

if __name__ == "__main__":
    main()

File: melvin_brain_monitor.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <unordered_map>
#include <chrono>
#include <thread>
#include <mutex>
#include <memory>
#include <algorithm>
#include <cstring>
#include <cstdint>
#include <set>
#include <optional>
#include <filesystem>
#include <iomanip>
#include <sstream>

#include "melvin_optimized_v2.h"

// ============================================================================
// BRAIN MONITORING SYSTEM
// ============================================================================

struct BrainActivity {
    uint64_t timestamp;
    uint64_t node_id;
    std::string activity_type;
    std::string details;
    double value;
    
    BrainActivity() : timestamp(0), node_id(0), value(0.0) {}
    
    BrainActivity(uint64_t ts, uint64_t id, const std::string& type, 
                  const std::string& det, double val)
        : timestamp(ts), node_id(id), activity_type(type), details(det), value(val) {}
};

struct BrainStats {
    uint64_t total_nodes;
    uint64_t total_connections;
    uint64_t hebbian_events;
    uint64_t pruning_events;
    uint64_t storage_bytes;
    double storage_mb;
    uint64_t uptime_seconds;
    uint64_t start_time;
    
    // Cognitive processing stats
    uint64_t cognitive_processing_events;
    uint64_t activation_clusters_formed;
    uint64_t interpretation_clusters_created;
    uint64_t candidate_responses_generated;
    uint64_t context_bias_applications;
    
    BrainStats() : total_nodes(0), total_connections(0), hebbian_events(0),
                   pruning_events(0), storage_bytes(0), storage_mb(0.0),
                   uptime_seconds(0), start_time(0), cognitive_processing_events(0),
                   activation_clusters_formed(0), interpretation_clusters_created(0),
                   candidate_responses_generated(0), context_bias_applications(0) {}
};

class MelvinBrainMonitor {
private:
    std::unique_ptr<MelvinOptimizedV2> melvin;
    std::vector<BrainActivity> activity_log;
    std::map<uint64_t, uint64_t> node_creation_times;
    std::map<uint64_t, uint64_t> connection_formation_times;
    std::vector<std::pair<uint64_t, uint64_t>> hebbian_events;
    std::vector<uint64_t> pruning_events;
    std::vector<BrainStats> stats_history;
    
    std::mutex monitor_mutex;
    std::thread monitoring_thread;
    bool monitoring_active;
    uint64_t monitor_start_time;
    
    static constexpr size_t MAX_ACTIVITY_LOG = 10000;
    static constexpr size_t MAX_STATS_HISTORY = 1000;
    static constexpr uint64_t MONITOR_INTERVAL_MS = 1000; // 1 second
    
public:
    MelvinBrainMonitor(const std::string& storage_path = "melvin_binary_memory") {
        melvin = std::make_unique<MelvinOptimizedV2>(storage_path);
        monitoring_active = false;
        monitor_start_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
        
        std::cout << "🧠 Melvin Brain Monitor initialized" << std::endl;
    }
    
    ~MelvinBrainMonitor() {
        stop_monitoring();
    }
    
    void start_monitoring() {
        if (monitoring_active) {
            std::cout << "⚠️  Monitoring already active" << std::endl;
            return;
        }
        
        monitoring_active = true;
        monitoring_thread = std::thread(&MelvinBrainMonitor::monitoring_loop, this);
        
        std::cout << "🔍 Brain monitoring started" << std::endl;
    }
    
    void stop_monitoring() {
        if (!monitoring_active) return;
        
        monitoring_active = false;
        if (monitoring_thread.joinable()) {
            monitoring_thread.join();
        }
        
        std::cout << "⏹️  Brain monitoring stopped" << std::endl;
    }
    
    void log_activity(uint64_t node_id, const std::string& activity_type, 
                     const std::string& details, double value = 0.0) {
        std::lock_guard<std::mutex> lock(monitor_mutex);
        
        uint64_t current_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
        
        BrainActivity activity(current_time, node_id, activity_type, details, value);
        activity_log.push_back(activity);
        
        // Keep log size manageable
        if (activity_log.size() > MAX_ACTIVITY_LOG) {
            activity_log.erase(activity_log.begin());
        }
        
        // Log to console
        std::cout << "📊 [" << format_timestamp(current_time) << "] "
                  << activity_type << " - Node " << std::hex << node_id 
                  << ": " << details << std::endl;
    }
    
    void log_node_creation(uint64_t node_id) {
        uint64_t current_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
        
        node_creation_times[node_id] = current_time;
        log_activity(node_id, "NODE_CREATED", "New node created");
    }
    
    void log_connection_formation(uint64_t source_id, uint64_t target_id) {
        uint64_t current_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
        
        uint64_t connection_id = std::hash<std::string>{}(
            std::to_string(source_id) + std::to_string(target_id));
        connection_formation_times[connection_id] = current_time;
        
        std::string details = "Connection " + std::to_string(source_id) + " -> " + std::to_string(target_id);
        log_activity(source_id, "CONNECTION_FORMED", details);
        
        hebbian_events.push_back({source_id, target_id});
    }
    
    void log_hebbian_event(uint64_t source_id, uint64_t target_id) {
        hebbian_events.push_back({source_id, target_id});
        std::string details = "Hebbian learning: " + std::to_string(source_id) + " -> " + std::to_string(target_id);
        log_activity(source_id, "HEBBIAN_LEARNING", details);
    }
    
    void log_pruning_event(uint64_t node_id) {
        pruning_events.push_back(node_id);
        log_activity(node_id, "NODE_PRUNED", "Node pruned due to low importance");
    }
    
    // Cognitive processing event logging
    void log_cognitive_processing_event(const std::string& input_text, uint64_t activated_nodes_count) {
        uint64_t current_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
        
        std::string details = "Processed input: " + input_text.substr(0, 30) + 
                             "... (" + std::to_string(activated_nodes_count) + " nodes activated)";
        log_activity(0, "COGNITIVE_PROCESSING", details, activated_nodes_count);
    }
    
    void log_activation_cluster_formation(uint64_t cluster_size) {
        log_activity(0, "ACTIVATION_CLUSTER", "Formed activation cluster with " + 
                     std::to_string(cluster_size) + " nodes", cluster_size);
    }
    
    void log_interpretation_cluster_creation(uint64_t cluster_id, float confidence) {
        log_activity(cluster_id, "INTERPRETATION_CLUSTER", 
                     "Created interpretation cluster with confidence " + std::to_string(confidence), 
                     confidence);
    }
    
    void log_candidate_response_generation(uint64_t response_id, float confidence) {
        log_activity(response_id, "CANDIDATE_RESPONSE", 
                     "Generated candidate response with confidence " + std::to_string(confidence), 
                     confidence);
    }
    
    void log_context_bias_application(uint64_t node_id, float bias_strength) {
        log_activity(node_id, "CONTEXT_BIAS", 
                     "Applied context bias with strength " + std::to_string(bias_strength), 
                     bias_strength);
    }
    
    BrainStats get_current_stats() {
        BrainStats stats;
        
        // Get stats from Melvin
        auto state = melvin->get_unified_state();
        
        stats.total_nodes = state.global_memory.total_nodes;
        stats.total_connections = state.global_memory.total_edges;
        stats.storage_mb = state.global_memory.storage_used_mb;
        stats.storage_bytes = static_cast<uint64_t>(stats.storage_mb * 1024 * 1024);
        stats.hebbian_events = state.global_memory.stats.hebbian_updates;
        stats.pruning_events = pruning_events.size();
        stats.uptime_seconds = state.system.uptime_seconds;
        stats.start_time = monitor_start_time;
        
        // Count cognitive processing events from activity log
        stats.cognitive_processing_events = 0;
        stats.activation_clusters_formed = 0;
        stats.interpretation_clusters_created = 0;
        stats.candidate_responses_generated = 0;
        stats.context_bias_applications = 0;
        
        for (const auto& activity : activity_log) {
            if (activity.activity_type == "COGNITIVE_PROCESSING") {
                stats.cognitive_processing_events++;
            } else if (activity.activity_type == "ACTIVATION_CLUSTER") {
                stats.activation_clusters_formed++;
            } else if (activity.activity_type == "INTERPRETATION_CLUSTER") {
                stats.interpretation_clusters_created++;
            } else if (activity.activity_type == "CANDIDATE_RESPONSE") {
                stats.candidate_responses_generated++;
            } else if (activity.activity_type == "CONTEXT_BIAS") {
                stats.context_bias_applications++;
            }
        }
        
        return stats;
    }
    
    void save_activity_report(const std::string& filename = "melvin_brain_report.json") {
        std::lock_guard<std::mutex> lock(monitor_mutex);
        
        std::ofstream file(filename);
        if (!file) {
            std::cerr << "❌ Failed to open report file: " << filename << std::endl;
            return;
        }
        
        auto stats = get_current_stats();
        
        file << "{\n";
        file << "  \"report_timestamp\": " << static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count()) << ",\n";
        file << "  \"monitoring_duration_seconds\": " << stats.uptime_seconds << ",\n";
        file << "  \"brain_stats\": {\n";
        file << "    \"total_nodes\": " << stats.total_nodes << ",\n";
        file << "    \"total_connections\": " << stats.total_connections << ",\n";
        file << "    \"hebbian_events\": " << stats.hebbian_events << ",\n";
        file << "    \"pruning_events\": " << stats.pruning_events << ",\n";
        file << "    \"storage_bytes\": " << stats.storage_bytes << ",\n";
        file << "    \"storage_mb\": " << std::fixed << std::setprecision(2) << stats.storage_mb << ",\n";
        file << "    \"cognitive_processing_events\": " << stats.cognitive_processing_events << ",\n";
        file << "    \"activation_clusters_formed\": " << stats.activation_clusters_formed << ",\n";
        file << "    \"interpretation_clusters_created\": " << stats.interpretation_clusters_created << ",\n";
        file << "    \"candidate_responses_generated\": " << stats.candidate_responses_generated << ",\n";
        file << "    \"context_bias_applications\": " << stats.context_bias_applications << "\n";
        file << "  },\n";
        file << "  \"recent_activity\": [\n";
        
        // Write recent activity (last 100 entries)
        size_t start_idx = (activity_log.size() > 100) ? activity_log.size() - 100 : 0;
        for (size_t i = start_idx; i < activity_log.size(); ++i) {
            const auto& activity = activity_log[i];
            file << "    {\n";
            file << "      \"timestamp\": " << activity.timestamp << ",\n";
            file << "      \"node_id\": \"" << std::hex << activity.node_id << "\",\n";
            file << "      \"activity_type\": \"" << activity.activity_type << "\",\n";
            file << "      \"details\": \"" << activity.details << "\",\n";
            file << "      \"value\": " << activity.value << "\n";
            file << "    }";
            if (i < activity_log.size() - 1) file << ",";
            file << "\n";
        }
        
        file << "  ]\n";
        file << "}\n";
        
        std::cout << "📄 Activity report saved to: " << filename << std::endl;
    }
    
    void print_real_time_stats() {
        auto stats = get_current_stats();
        
        std::cout << "\n🧠 MELVIN BRAIN STATS" << std::endl;
        std::cout << "=====================" << std::endl;
        std::cout << "📊 Nodes: " << stats.total_nodes << std::endl;
        std::cout << "🔗 Connections: " << stats.total_connections << std::endl;
        std::cout << "⚡ Hebbian Events: " << stats.hebbian_events << std::endl;
        std::cout << "🗑️ Pruning Events: " << stats.pruning_events << std::endl;
        std::cout << "💾 Storage: " << std::fixed << std::setprecision(2) << stats.storage_mb << " MB" << std::endl;
        std::cout << "⏱️ Uptime: " << format_duration(stats.uptime_seconds) << std::endl;
        std::cout << "📈 Activity Log Entries: " << activity_log.size() << std::endl;
        std::cout << "\n🧠 Cognitive Processing Stats:" << std::endl;
        std::cout << "🔍 Cognitive Events: " << stats.cognitive_processing_events << std::endl;
        std::cout << "🎯 Activation Clusters: " << stats.activation_clusters_formed << std::endl;
        std::cout << "💭 Interpretation Clusters: " << stats.interpretation_clusters_created << std::endl;
        std::cout << "💬 Candidate Responses: " << stats.candidate_responses_generated << std::endl;
        std::cout << "🎛️ Context Bias Applications: " << stats.context_bias_applications << std::endl;
        std::cout << "=====================" << std::endl;
    }
    
private:
    void monitoring_loop() {
        while (monitoring_active) {
            // Capture current stats
            auto stats = get_current_stats();
            stats_history.push_back(stats);
            
            // Keep history size manageable
            if (stats_history.size() > MAX_STATS_HISTORY) {
                stats_history.erase(stats_history.begin());
            }
            
            // Print stats every 10 seconds
            static uint64_t last_print = 0;
            uint64_t current_time = static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::seconds>(
                std::chrono::system_clock::now().time_since_epoch()).count());
            
            if (current_time - last_print >= 10) {
                print_real_time_stats();
                last_print = current_time;
            }
            
            // Sleep for monitoring interval
            std::this_thread::sleep_for(std::chrono::milliseconds(MONITOR_INTERVAL_MS));
        }
    }
    
    std::string format_timestamp(uint64_t timestamp) {
        auto time_point = std::chrono::system_clock::from_time_t(timestamp);
        auto time_t = std::chrono::system_clock::to_time_t(time_point);
        auto tm = *std::localtime(&time_t);
        
        std::ostringstream oss;
        oss << std::put_time(&tm, "%H:%M:%S");
        return oss.str();
    }
    
    std::string format_duration(uint64_t seconds) {
        uint64_t hours = seconds / 3600;
        uint64_t minutes = (seconds % 3600) / 60;
        uint64_t secs = seconds % 60;
        
        std::ostringstream oss;
        if (hours > 0) {
            oss << hours << "h " << minutes << "m " << secs << "s";
        } else if (minutes > 0) {
            oss << minutes << "m " << secs << "s";
        } else {
            oss << secs << "s";
        }
        return oss.str();
    }
};

// ============================================================================
// MAIN FUNCTION
// ============================================================================

int main() {
    std::cout << "🧠 MELVIN BRAIN MONITOR (C++)" << std::endl;
    std::cout << "=============================" << std::endl;
    
    try {
        // Initialize brain monitor
        MelvinBrainMonitor monitor;
        
        // Start monitoring
        monitor.start_monitoring();
        
        // Simulate some brain activity
        std::cout << "🧪 Simulating brain activity..." << std::endl;
        
        // Process some test inputs
        auto melvin = std::make_unique<MelvinOptimizedV2>();
        
        uint64_t text_id = melvin->process_text_input("Hello, this is a test of the brain monitor!");
        monitor.log_node_creation(text_id);
        
        uint64_t code_id = melvin->process_code_input("def test_function():\n    return 'Hello, World!'");
        monitor.log_node_creation(code_id);
        
        // Simulate some connections
        monitor.log_connection_formation(text_id, code_id);
        monitor.log_hebbian_event(text_id, code_id);
        
        // Simulate cognitive processing
        std::cout << "🧠 Simulating cognitive processing..." << std::endl;
        std::string cognitive_input = "How does machine learning work?";
        auto cognitive_result = melvin->process_cognitive_input(cognitive_input);
        
        monitor.log_cognitive_processing_event(cognitive_input, cognitive_result.activated_nodes.size());
        monitor.log_activation_cluster_formation(cognitive_result.activated_nodes.size());
        
        for (size_t i = 0; i < cognitive_result.clusters.size(); ++i) {
            monitor.log_interpretation_cluster_creation(i, cognitive_result.clusters[i].confidence);
        }
        
        monitor.log_candidate_response_generation(0, cognitive_result.confidence);
        
        // Let monitoring run for a bit
        std::this_thread::sleep_for(std::chrono::seconds(5));
        
        // Print final stats
        monitor.print_real_time_stats();
        
        // Save report
        monitor.save_activity_report();
        
        // Stop monitoring
        monitor.stop_monitoring();
        
        std::cout << "\n🎉 Brain monitoring test completed successfully!" << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "❌ Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}
